src/tabular-api-surface.ts

```diff
@@ -547,16 +586,33 @@ Please use the format 'prezzy' or '${instance.name}/tables/prezzy'.`);
               //
               error.code = grpc.status.CANCELLED;
             }
+            metricsCollector?.onAttemptComplete(
+              this.bigtable.projectId,
+              error.code
+            );
+            metricsCollector?.onOperationComplete(
+              this.bigtable.projectId,
+              error.code
+            );
             userStream.emit('error', error);
           }
         })
         .on('data', _ => {
           // Reset error count after a successful read so the backoff
           // time won't keep increasing when as stream had multiple errors
           numConsecutiveErrors = 0;
+          metricsCollector?.onResponse(this.bigtable.projectId);
         })
         .on('end', () => {
           activeRequestStream = null;
+          metricsCollector?.onAttemptComplete(
+            this.bigtable.projectId,
+            grpc.status.OK
+          );
```

Comment:
FWIW, in Python I created a separate `onOperationSucceeded` helper to call onAttemptComplete and onOpeationComplete with grpc.status.OK. This makes it a bit easier to see the happy-path, and can simplify things a bit. Up to you if thats helpful here

Reply:
I added a onLastAttemptCompleted method which simplifies things in a few places. It will reduce the amount of code necessary for methods that use the metrics collector.